File: java/src/main/java/com/example/cote/PGMRS/키패드_누르기.java
package com.example.cote.PGMRS;

import java.util.*;
import java.io.*;
import java.nio.file.*;
import static java.util.stream.Collectors.*;

public class 키패드_누르기 {

	public static class Hand {
		private final int baseX;
		public final String hand;
		public final float preference;
		private int x;
		private int y;

		public Hand(String hand, boolean isPreferred, int x) {
			this.hand = hand;
			this.baseX = x;
			this.preference = isPreferred ? 0.5f : 0;
			this.x = x;
			this.y = 3;
		}

		public float distance(int x, int y) {
			if (x == baseX) return 0;
			int distance = Math.abs(x - this.x) + Math.abs(y - this.y);
			return distance - preference;
		}

		public Hand move(int x, int y) {
			this.x = x;
			this.y = y;

			return this;
		}
	}

	private static int getX(int number) {
		if (number == 0) return 1;
		return (number - 1) % 3;
	}

	private static int getY(int number) {
		if (number == 0) return 3;
		return (number - 1) / 3;
	}

	private static String press (int number, Hand right, Hand left) {
		int x = getX(number);
		int y = getY(number);

		float rDistance = right.distance(x, y);
		float lDistance = left.distance(x, y);

		if (lDistance < rDistance) {
			return left.move(x,y).hand;
		}
		return right.move(x,y).hand;
	}

	private static String solve(int[] numbers, String hand) {
		Hand right = new Hand("R", hand.equals("right"), 2);
		Hand left = new Hand("L", hand.equals("left"), 0);

		return Arrays.stream(numbers)
			.mapToObj(n -> press(n, right, left))
			.collect(joining());
	}

	public static void main(String... args) throws IOException {

		Path p = Paths.get(System.getProperty("user.dir")+"/data/키패드_누르기.txt");
		BufferedReader reader = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(reader.readLine());

		while (testCases > 0) {

			int[] numbers = Arrays.stream(reader.readLine().split(" "))
				.mapToInt(Integer::parseInt)
				.toArray();

			String hand = reader.readLine().trim();

			String answer = reader.readLine();

			System.out.println("Answer : "+ answer);
			System.out.println("Solved : "+ solve(numbers, hand));

			testCases--;
		}
	}
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/키패드_누르기.kt
package com.example.cote.PGMRS;

import java.io.*
import java.nio.file.*

class Hand(val hand: String,
	var y : Int,
	var x : Int,
	val baseX: Int,
	preferedHand: String) {
	
	val prefered: Boolean
	
	init {
		prefered = preferedHand.equals(hand)
	}
	
	fun distance(y: Int, x: Int): Double 
		= when {
			x == baseX -> 0.0
			prefered -> Math.abs(this.x - x) + Math.abs(this.y - y) - 0.5
			else ->  (Math.abs(this.x - x) + Math.abs(this.y - y)).toDouble()
		}

	fun move(y: Int, x: Int): Hand {
		this.x = x;
		this.y = y;
		return this;
	}
}

fun getY(number: Int): Int 
	= if (number == 0) 3 else (number - 1) / 3

fun getX(number: Int): Int 
	= if (number == 0) 1 else (number - 1) % 3 

fun press(number: Int, left: Hand, right: Hand): String {
	val y = getY(number)
	val x = getX(number)

	val lDistance = left.distance(y, x)
	val rDistance = right.distance(y, x)

	return if (lDistance > rDistance) right.move(y, x).hand
		else left.move(y, x).hand
}

fun solve(numbers: Array<Int>, prefered: String): String {
	var preferedHand  = if (prefered.equals("left")) "L" else "R"
	val left = Hand("L", 3, 0, 0, preferedHand)
	val right = Hand("R", 3, 2, 2, preferedHand)

	return numbers.map{
		press(it, left, right)
	}.joinToString("")
}

fun main() {

	val p = Paths.get(System.getProperty("user.dir") + "/data/키패드_누르기.txt")
	val reader = Files.newBufferedReader(p)

	repeat(reader.readLine().toInt()) {
		val numbers = reader.readLine().split(" ")
			.map(String::toInt)
			.toTypedArray()

		val prefered = reader.readLine()
		val answer = reader.readLine()

		println("$answer ${solve(numbers, prefered)}")
	}
}